lib/Transforms/TensorToScalars/TensorToScalars.cpp

```diff
@@ -0,0 +1,144 @@
+    for (int i = 0; i < dim; ++i) {
+      Value index = builder.create<arith::ConstantIndexOp>(loc, i);
+      Value element = builder.create<tensor::ExtractOp>(loc, input, index);
+      values.push_back(element);
+    }
+  }
+  return values;
+}
+
+class ConvertFromElementsOp
+    : public OneToNOpConversionPattern<tensor::FromElementsOp> {
+ public:
+  using OneToNOpConversionPattern<
+      tensor::FromElementsOp>::OneToNOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      tensor::FromElementsOp op, OpAdaptor adaptor,
+      OneToNPatternRewriter &rewriter) const override {
+    // This conversion only operates on tensors of static shape,
+    // but no check is necessary here as from_elements' shape is always static
+
+    // Replace the current op with the flattened operands.
+    // This should already match the "natural" order expected by this pass.
+    rewriter.replaceOp(op, adaptor.getFlatOperands(),
+                       adaptor.getResultMapping());
+    return success();
+  }
+};
+
+class ConvertInsertOp : public OneToNOpConversionPattern<tensor::InsertOp> {
+ public:
+  using OneToNOpConversionPattern<tensor::InsertOp>::OneToNOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      tensor::InsertOp op, OpAdaptor adaptor,
+      OneToNPatternRewriter &rewriter) const override {
+    // This conversion only operates on tensors of static shape
+    if (!op.getResult().getType().hasStaticShape()) return failure();
+
+    // We can only support statically known indices
+    // that have been constant-folded to a single arith.constant op
+    for (auto idx : op.getIndices()) {
+      if (!llvm::isa<arith::ConstantIndexOp>(idx.getDefiningOp()))
+        return failure();
+    }
+
+    // Compute the insertion offset (in dimension-by-dimension order):
+    int64_t multiplier = 1;
+    int64_t offset = 0;
+    for (auto [dim, idx] :
+         llvm::zip(op.getResult().getType().getShape(), op.getIndices())) {
+      offset +=
+          idx.getDefiningOp<arith::ConstantIndexOp>().value() * multiplier;
+      multiplier *= dim;
+    }
+
+    // get converted "tensor" operand from op (likely a unrealized_builtin_cast)
+    SmallVector<Value> elements = adaptor.getOperands()[1];
+    // replace element at offset with the "scalar" operand to be inserted
+    elements[offset] = adaptor.getOperands()[0].front();
+    // replace the current op with the converted operands.
+    rewriter.replaceOp(op, elements, adaptor.getResultMapping());
+    return success();
+  }
+};
+
+struct TensorToScalars : impl::TensorToScalarsBase<TensorToScalars> {
+  using TensorToScalarsBase::TensorToScalarsBase;
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+
+    OneToNTypeConverter typeConverter;
+    typeConverter.addConversion([](Type type) { return type; });
+    typeConverter.addConversion(
+        [](RankedTensorType tensorType,
+           SmallVectorImpl<Type> &types) -> std::optional<LogicalResult> {
```

Comment:
maybe in your blog post you can explain why one might return `std::nullopt` in this function :)

Reply:
Good catch, this should definitively "soft" fail